### PR TITLE
Fixes #12013: Fix for a pure virtual call in Sds constructor

### LINKED ISSUE #12013: Pure virtual call in SdsApi
Seems like it's possible to hit a pure virtual call when calling `Envoy::Secret::SecretManagerImpl::findOrCreateGenericSecretProvider` during filter construction with a file-based SDS secret:

```
#4  0x0000000004546e31 in std::terminate() ()
#5  0x000000000454785f in __cxa_pure_virtual ()
#6  0x000000000287b0d9 in Envoy::Secret::SdsApi::onConfigUpdate (this=0x515e610, resources=..., version_info=...) at source/common/secret/sds_api.cc:56
#7  0x000000000314e5f0 in Envoy::Config::FilesystemSubscriptionImpl::refresh (this=0x5836d20) at source/common/config/filesystem_subscription_impl.cc:59
#8  0x000000000314e3d0 in Envoy::Config::FilesystemSubscriptionImpl::start (this=0x5836d20) at source/common/config/filesystem_subscription_impl.cc:33
#9  0x000000000287bc5e in Envoy::Secret::SdsApi::initialize (this=0x515e610) at source/common/secret/sds_api.cc:115
#10 0x000000000287c668 in Envoy::Secret::SdsApi::SdsApi(envoy::config::core::v3::ConfigSource, absl::string_view, Envoy::Config::SubscriptionFactory&, Envoy::TimeSource&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Stats::Store&, Envoy::Init::Manager&, std::function<void ()>, Envoy::Event::Dispatcher&, Envoy::Api::Api&)::$_0::operator()() const (this=0x5851a78) at source/common/secret/sds_api.cc:24
```

Seems like the SdsApi is initialized and triggers the `init_manager_` callback immediately since we're past the startup sequence, which triggers start -> refresh on the file subscription inline. This results in `onConfigUpdate` being called, which invokes virtual functions that have not yet been initialized.

### PR DESCRIPTION
**Pull request by dmitri-d**
Signed-off-by: Dmitri Dolguikh <ddolguik@redhat.com>

Fixes https://github.com/envoyproxy/envoy/issues/12013

Risk Level: Low
Testing: added a test
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by dmitri-d** [comment]
ping @snowp

File: source/common/secret/sds_api.cc [hunk 1]
```diff
@@ -15,8 +15,8 @@ namespace Secret {
 SdsApi::SdsApi(envoy::config::core::v3::ConfigSource sds_config, absl::string_view sds_config_name,
                Config::SubscriptionFactory& subscription_factory, TimeSource& time_source,
                ProtobufMessage::ValidationVisitor& validation_visitor, Stats::Store& stats,
-               Init::Manager& init_manager, std::function<void()> destructor_cb,
-               Event::Dispatcher& dispatcher, Api::Api& api)
+               Init::Manager&, std::function<void()> destructor_cb, Event::Dispatcher& dispatcher,
```
Comment (snowp) [inline source/common/secret/sds_api.cc:18]: Remove this arg if its no longer used

File: source/common/secret/sds_api.cc [hunk 2]
```diff
@@ -30,11 +30,11 @@ SdsApi::SdsApi(envoy::config::core::v3::ConfigSource sds_config, absl::string_vi
   // This has to happen here (rather than in initialize()) as it can throw exceptions.
   subscription_ = subscription_factory_.subscriptionFromConfigSource(
       sds_config_, Grpc::Common::typeUrl(resource_name), stats_, *this, resource_decoder_);
+
   // TODO(JimmyCYJ): Implement chained_init_manager, so that multiple init_manager
```
Comment (snowp) [inline source/common/secret/sds_api.cc:34]: Is this TODO still relevant?

File: source/common/secret/sds_api.cc (hunk 1 above)
Comment (dmitri-d) [reply source/common/secret/sds_api.cc]: done

File: source/common/secret/sds_api.cc (hunk 2 above)
Comment (dmitri-d) [reply source/common/secret/sds_api.cc:33]: I think so; Added a note where to look re: handling of init targets

**Comment by dmitri-d** [comment]
* responded to feedback

**Comment by lizan** [comment]
fix clang_tidy (about unused `using ByMove` etc)

**Comment by dmitri-d** [comment]
* fixed clang-tidy errros

**Comment by dmitri-d** [comment]
> Does this avoid making any calls to virtual functions in the ctor or just bypasses the specific issue with the base class not being initialized?

The issue is due to virtual functions are being called from the base class (where these functions are pure virtual) ctor: at this point sub-class hasn't yet been instantiated, and these functions cannot be yet resolved. The issue in this case is resolved by moving the call to the init_manager to the concrete sub-class ctor -- this guarantees that the functions have implementations. 

I agree that this is still not ideal, but I think the bigger issue lies with component initalization/dependency wiring and init_manager.

**Comment by snowp** [comment]
How about extracting the Init::Target registration into a function that needs to be called during object creation instead of relying on virtual dispatch in the ctor? The proposed solution works for the current class hierarchy, but I'd be a bit concerned that future changes here could trip over this issue again (e.g. new `SdsApi` implementations are added with more complex class hierarchy). WDYT?

**Comment by dmitri-d** [comment]
The calls to virtual methods are a result of  Init::Target registration, but aren't direct calls -- they are made when the subscription eventually invokes the `onConfigUpdate` callback, which then invokes `validateConfig` and `setSecret` defined in concrete subclasses. 

I also realised the problem is more complicated than I thought: using `this` pointer should normally result in correct overloaded method being called. But likely b/c when the callback above is invoked, only the base class (and its vtable) have been constructed, the call is handled by the base class, which defined a pure virtual method only.

Postponing the call until we are in the subclass constructor seems to resolve the issue. The only other thing I can think of is to use a `Cleanup` class to delay this initialization until the ctor is completed.

**Comment by dmitri-d** [comment]
ping @snowp

**Comment by dmitri-d** [comment]
ping @snowp

**Comment by dmitri-d** [comment]
> This is why the style guide suggests never making virtual calls in ctor/dtors, because the vtable might be partially constructed/destructed. Using a Cleanup-like class in the base ctor probably wouldn't work, as I believe the base ctor will return before the ctor of the sub class. As such I don't think there is any way of doing virtual calls in a ctor that is safe against further sub classing, which is why I suggested handling the init logic as part of the static create method instead. I think we could ensure that this init call happens in test etc. by ASSERTing in the SdsApi dtor that initialization has been done, so now we get 1) no funny vtable business and 2) ASSERT guarantee'd initialization. WDYT?

I agree with pretty much everything above. My concern is that by delaying the addition of SdsApi init_target to the init_manager until after the concrete class has been instantiated we are breaking a convention (I think only VHDS doesn't follow it, but it's weird anyway):  on return from ctor we get a subscription instance that will be initialised/is initialised without any additional work. It's also somewhat subtle and easy to miss when looking at the code. Moving the logic into concrete class ctors is more obvious and harder to miss (also: is this class going to be extended further?). Personally, I'd rather keep the logic in concrete class ctors, but if you feel strongly about this, I can move it out to `create` methods in concrete implementations of SdsApi.

**Comment by dmitri-d** [comment]
* fixed formatting
* merged in latest mater

**Comment by dmitri-d** [comment]
ping @snowp

File: source/common/secret/sds_api.h [hunk 3]
```diff
@@ -45,11 +45,12 @@ class SdsApi : public Envoy::Config::SubscriptionBase<
   SdsApi(envoy::config::core::v3::ConfigSource sds_config, absl::string_view sds_config_name,
          Config::SubscriptionFactory& subscription_factory, TimeSource& time_source,
          ProtobufMessage::ValidationVisitor& validation_visitor, Stats::Store& stats,
-         Init::Manager& init_manager, std::function<void()> destructor_cb,
-         Event::Dispatcher& dispatcher, Api::Api& api);
+         std::function<void()> destructor_cb, Event::Dispatcher& dispatcher, Api::Api& api);
 
   SecretData secretData();
 
+  void registerInitTarget(Init::Manager& init_manager) { init_manager.add(init_target_); }
```
Comment (snowp) [inline source/common/secret/sds_api.h:52]: Could you add a bool `initialized_` that you set here and ASSERT on in the dtor so we're ensured that this is called?

File: source/common/secret/sds_api.cc [hunk 4]
```diff
@@ -30,11 +29,14 @@ SdsApi::SdsApi(envoy::config::core::v3::ConfigSource sds_config, absl::string_vi
...
   // can be chained together to behave as one init_manager. In that way, we let
   // two listeners which share same SdsApi to register at separate init managers, and
   // each init manager has a chance to initialize its targets.
-  init_manager.add(init_target_);
+  //
+  // Note that addition of initialization targets to init manager is handled in constructors SdsApi
```
Comment (snowp) [inline source/common/secret/sds_api.cc:38]: not relevant anymore?

**Comment by dmitri-d** [comment]
* added an assert in dtor to confirm that the init_target has been registered with the init_manager.

I think a better place for this flag is in the init_target itself. Such a change would expand the scope of this PR though. I can create an issue so it doesn't get lost.

**Comment by dmitri-d** [comment]
* merged in latest master

**Comment by snowp** [comment]
>I think a better place for this flag is in the init_target itself. Such a change would expand the scope of this PR though. I can create an issue so it doesn't get lost.

Yep I agree

**Comment by snowp** [comment]
This LGTM, can you take a look at CI?

**Comment by dmitri-d** [comment]
* Fixed an issue reported by clang-tidy

**Comment by dmitri-d** [comment]
* merged in latest master

**Comment by dmitri-d** [comment]
ping @snowp

**Pull request merged**